src/components/Checkins/GoogleMap.tsx

```diff
@@ -0,0 +1,228 @@
+import React, { Component } from 'react';
+import styled from 'styled-components';
+import { makeTimeString } from '../../business/time';
+import Checkin from '../../models/Checkin';
+import CheckinDetails from '../../models/CheckinDetails';
+
+const unselectedMarkerOpacity = 0.5;
+
+const Map = styled.div`
+    height: 100%;
+`;
+
+interface GoogleMapProps {
+    checkins: Checkin[];
+    setSelected(details: CheckinDetails): void;
+}
+
+interface GoogleMapState {
+    hasGoogleLoaded: boolean;
+    mapRef: React.RefObject<HTMLDivElement>;
+}
+
+type PlacesService = google.maps.places.PlacesService;
+type PlaceResult = google.maps.places.PlaceResult;
+type Marker = google.maps.Marker;
+
+class GoogleMap extends Component<GoogleMapProps, GoogleMapState> {
+    private map?: google.maps.Map = undefined;
+    private markers: Marker[] = [];
+    private hasLoadedMarkers = false;
+
+    constructor(props: GoogleMapProps) {
+        super(props);
+        window.googleInit = this.googleInit;
+        this.state = {
+            hasGoogleLoaded: window.hasGoogleLoaded,
+            mapRef: React.createRef<HTMLDivElement>(),
+        };
+    }
+
+    public render = () => {
+        return <Map ref={this.state.mapRef}>Loading map...</Map>;
+    }
+
+    public componentDidUpdate = () => {
+        this.createMapIfNecessary();
+        this.updateMapIfNecessary();
+    }
+
+    public componentDidMount = () => {
+        this.createMapIfNecessary();
+        this.updateMapIfNecessary();
+    }
+
+    // MARK: - Private
+
+    private googleInit = () => {
+        console.log('google init in component');
+        window.hasGoogleLoaded = true;
+        this.setState({ hasGoogleLoaded: true });
+    }
+
+    private createMapIfNecessary = () => {
+        // bounce if google is not initialized
+        if (!this.state.hasGoogleLoaded || this.map) { return; }
+        console.log('creating map...');
+
+        const node = this.state.mapRef.current;
+        if (node) {
+            this.map = new google.maps.Map(node);
+        }
+    }
+
+    private updateMapIfNecessary = () => {
+        // bounce if map or google doesn't exist
+        if (!this.map || !this.state.hasGoogleLoaded || this.hasLoadedMarkers) { return; }
+        console.log('updating map...');
+        this.hasLoadedMarkers = true;
+        this.markers.forEach((m) => m.setMap(null));
+        this.markers = [];
+
+        const { checkins } = this.props;
+        const service = new google.maps.places.PlacesService(this.map);
+        this.recursiveFetchPlace(0, Number.MAX_SAFE_INTEGER, checkins, service);
+    }
+
+    private recursiveFetchPlace = (index: number, selectedIndex: number, checkins: Checkin[], service: PlacesService) => {
```

Comment:
This is more of a style thing, but if you were to mark this as an [async function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function), e.g.:
```typescript
private recursiveFetchPlace = async (index: number, selectedIndex: number, checkins: Checkin[], service: PlacesService) => {
```
you could make this look like:
```typescript
const placeId = await this.fetchPlaceId(checkin, service);
const details = await this.fetchPlaceDetail(checkin, service, placeid);
// ...
```